### ISSUE TITLE
(Shared) Formula recalculation performance regression

### ISSUE THREAD

**Issue by Nextra** [ClosedXML/ClosedXML#2350]
**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature
- [ ] Question

**Did you test against the latest CI build?**
- [x] Yes
- [ ] No

**Version of ClosedXML**

`v0.102.2`, `v0.104.0`

**What is the current behavior?**

Shared formulas in the example file recalculate slowly/infinitely.

**What is the expected behavior or new feature?**

Shared formulas in the example file recalculate reasonably quickly.

**Is this a regression from the previous version?**

Yes. I found this issue in a more complex file, where the regression causes recalculation with version `0.102.2` to never complete. Versions `0.100.3` and below process this file in around one second, while `0.103.0` and newer take well over a minute.

The example file I attached is processed as follows on my machine:
|           Version | Recalculation Behavior        |
| ----------------: | :---------------------------- |
|       `<=0.100.3` | ~1s to complete               |
| `0.101.0-0.102.2` | Never completes, memory spike |
|       `>=0.103.0` | ~12s to complete              |

I couldn't quite reproduce the 60-80x discrepancy with this simplified example, but it hopefully illustrates the problem well enough.

## Reproducibility

Anything that causes recalculations of cells in column B is enough to trigger the issue, but for brevity I just trigger a full recalculation:

```c#
using ClosedXML.Excel;

var path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "SharedFormula.xlsx");

var workbook = new XLWorkbook(path);
var sheet = workbook.Worksheet("Sheet1");

sheet.RecalculateAllFormulas();
```

As mentioned in the file, versions 101 and 102 will not manage to recalculate Sheet 2, even thought it contains just three rows.

When the formulas in column B are not shared, versions 101 and 102 will not hit the regression and perform the same as 100. This can be verified by regenerating a functionally equivalent file via ClosedXML itself:

```c#
using ClosedXML.Excel;

var workbook = new XLWorkbook();
var sheet = workbook.AddWorksheet("Sheet1");

for (var i = 1; i <= 2000; ++i) {
    sheet.Cell($"A{i}").Value = $"Lookup {i}";
    sheet.Cell($"B{i}").FormulaA1 = $"=\"Value \"&C{i}&\"\"";
    sheet.Cell($"C{i}").FormulaA1 = $"=VLOOKUP($A{i},$A:$D,4,FALSE)";
    sheet.Cell($"D{i}").Value = i;
}

sheet.RecalculateAllFormulas();
```

Version 104 however will still be slower than the other versions.

- [x] I attached a sample spreadsheet.  (You can drag files on to this issue)
[SharedFormula.xlsx](https://github.com/ClosedXML/ClosedXML/files/15495801/SharedFormula.xlsx)

**Comment by ThrowingClubs** [comment]
I came here to write an issue about poor performance with VLOOKUP, but I think my issue is related to this. 

As best I can tell the issue lies in the XLCalcEngine's TryGetPrecedentAreas functionality.  In my case the end range of one of the precedent areas is getting set to the max cell (1048576) even when there are only a few rows filled in with data. I don't know exactly what the knock on effects are but it's pretty easy to guess that an extra million cells to search through would be the cause of the performance concern.

Also, when I downgrade to version 100 I don't have the same issue.

**Comment by Nextra** [comment]
My specific issue is all but fixed with the release version of 0.104.1. 

The file I discovered this issue in is still processed slightly slower than 0.100.3, but only by ~10%. This is enough for my use case.
